features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/net/ResourceRequestBodyFactory.kt

```diff
@@ -0,0 +1,147 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.net
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.api.storage.RawBatchEvent
+import com.datadog.android.sessionreplay.internal.exception.InvalidPayloadFormatException
+import com.datadog.android.sessionreplay.internal.utils.MiscUtils
+import com.google.gson.JsonObject
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.MultipartBody
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import kotlin.jvm.Throws
+
+internal class ResourceRequestBodyFactory(
+    private val internalLogger: InternalLogger
+) {
+
+    @Throws(InvalidPayloadFormatException::class)
+    internal fun create(
+        resources: List<RawBatchEvent>
+    ): RequestBody {
+        val deserializedResources: List<ResourceEvent> = resources.mapNotNull {
+            val resourceMetadata = MiscUtils.safeDeserializeToJsonObject(internalLogger, it.metadata)
+
+            if (resourceMetadata != null) {
+                val applicationId = MiscUtils.safeGetStringFromJsonObject(
+                    internalLogger,
+                    resourceMetadata,
+                    APPLICATION_ID_KEY
+                )
+                val filename = MiscUtils.safeGetStringFromJsonObject(
+                    internalLogger,
+                    resourceMetadata,
+                    FILENAME_KEY
+                )
+
+                if (applicationId != null && filename != null) {
+                    ResourceEvent(
+                        applicationId = applicationId,
+                        identifier = filename,
+                        it.data
+                    )
+                } else {
+                    null
+                }
+            } else {
+                null
+            }
+        }
+
+        if (deserializedResources.isEmpty()) {
+            @Suppress("ThrowingInternalException")
+            throw InvalidPayloadFormatException(NO_RESOURCES_TO_SEND_ERROR)
+        }
+
+        val applicationId = getApplicationId(deserializedResources)
+
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
```

Comment:
> [!WARNING]
> It might be handled by the caller in the current implementation right now, but you can't know for sure it will be in the future. Again this could lead to future unforeseen crashes